docs/proposals/signal-api.md

```diff
@@ -0,0 +1,172 @@
+The signal notification will be in fact a mapping from the notification name to the actual POSIX name. In example,
+user may want to define notification called *bump*, give it a *signal* type and point it to the SIGHUP signal.
+After creating such a pod, he would be able to send it a *bump* notification, which will result.
+
+The idea of the notifications allows to create many types, but only implementing the signal type is in the scope
+of this proposal.
+
+Once me make agreement on the API part, the document will be extended with the implementation details.
+
+One thing that needs to be noticed is that restarting scenarios are not a goad of this proposal. The only goal here
+is to send single signals, no more, no less. More complicated restart scenarios (like sending SIGTERM, then waiting
+for graceful termination, then sending SIGKILL after reaching timeout) may need another resource in Kubernetes API,
+which will not expose any POSIX terminology to the user.
+
+## Motivation
+
+Many Kubernetes users and operators would like to send custom signals to the processes running in containers. That
+may be motivated both by default behaviour that signals are bringing and the custom signal handling which the
+program is doing.
+
+The more concrete use-cases are described below.
+
+## Use-Cases
+
+#### Re-reading configuration
+
+Many applications follow the convention to re-read configuration files when handling the SIGHUP signal. Ability
+to send SIGHUP signal will allow us to reconfigure pods without updating their definition or removing them. Also,
+we can consider using this API in the watch mechanism in kubelet to reconfigure a pod automatically on the ConfigMap
+change.
+
+## Analysis
+
+The system needs to be able to:
+
+1. Model correctly which signals are supported
+2. Allow to define what signals can be used for each pod (and whether they can be used at all)
+3. Provide custom names for the chosen signals in pod
+
+## Proposed design
+
+Signal API will be a part of pod API. There will be no separate API group for that.
+
+#### Schema
+
+The proposed schema is as follow.
+
+There will be a type Notification being a part of the Container (inside the PodSpec):
+
+```go
+type Notification struct {
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Type string `json:"type" protobuf:"bytes,2,opt,name=type"`
+	Signal string `json:"signal" protobuf:"bytes,3,opt,name=signal"`
+}
+```
+
+The *signal* parameter will be able to take any name of POSIX signal and that will be a criterion to validate that
+field.
+
+And there will be an API subresource of `pod` called `notify`, with the options:
+- name
+
+#### Endpoints
+
+There will be PUT and POST endpoint, rooted at /api/v1/, with the parameters as listed:
+- /namespaces/{namespace}/pods/{pod}/notify
+  - name  - name of the notification to send
+
+#### ContainerRuntime interface
```

Comment:
To be clear: What I proposed would set a *significant* precedent in the
API, so it must be entered into VERY carefully.  I wonder if we can find a
way to let you prototype a solution without blocking on that.

What if, for the sake of proving the idea, you added an `
pod.alpha.kubernetes.io/pending-notificiations` annotation, whose values
was a JSON-encoded list of strings.  You can do all the work to make
kubelet react to that and to make apiserver convert .../notify REST ops
into that, without investing a ton of time into an idea that needs big
design.

Once we have working proof, it's more likely that we can drive some
agreement about whether to do imperatives, or maybe just clean-up and keep
it as an annotation.

On Thu, Jul 7, 2016 at 6:34 AM, Michal Rostecki <notifications@github.com>
wrote:

> In docs/proposals/signal-api.md
> <https://github.com/kubernetes/kubernetes/pull/26884#discussion_r69906427>
> :
>
> > +}
> > +```
> > +
> > +The *signal* parameter will be able to take any name of POSIX signal and that will be a criterion to validate that
> > +field.
> > +
> > +And there will be an API subresource of `pod` called `notify`, with the options:
> > +- name
> > +
> > +#### Endpoints
> > +
> > +There will be PUT and POST endpoint, rooted at /api/v1/, with the parameters as listed:
> > +- /namespaces/{namespace}/pods/{pod}/notify
> > +  - name  - name of the notification to send
> > +
> > +#### ContainerRuntime interface
>
> @thockin <https://github.com/thockin> I like the idea of this
> PodOperation struct. From what I understand, your idea assumes that the
> PodOperation will be applied by some pod worker in kubelet?
>
> I'll try to include that idea in my proposal soon, but at the same time, I
> have to dig in kubelet pod workers code to see how can I exactly implement
> that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26884/files/f64c269788600b6847cc66e6271293f9c77bc2fa#r69906427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFVgVMQU_4tfsgsLLeJZ0wxrrtPg7LsPks5qTQB3gaJpZM4IuvkT>
> .
>